### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
# IMPLEMENTACIÓN ORIGINAL

import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.034317970275878906 seconds
34.5 ms ± 378 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [2]:
# 2. Usando listas de Python

# a) Crear lista con range
lista = list(range(value))
print(f"Lista creada con {len(lista)} elementos")

# b) Sumar con bucle for
print("\n--- Suma usando bucle for ---")
start_time = time.time()
suma_for = 0
for num in lista:
    suma_for += num
end_time = time.time()
print(f"Tiempo con bucle for: {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_for}")

# c) Sumar con función sum()
print("\n--- Suma usando función sum() ---")
start_time = time.time()
suma_func = sum(lista)
end_time = time.time()
print(f"Tiempo con función sum(): {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_func}")

# Comparar con timeit
print("\n--- Comparación con %timeit ---")
print("Bucle for:")
%timeit -r 2 -n 1 suma_for = 0; [suma_for := suma_for + num for num in lista]

print("\nFunción sum():")
%timeit -r 2 -n 1 sum(lista)

Lista creada con 1000000 elementos

--- Suma usando bucle for ---
Tiempo con bucle for: 0.0760 segundos
Resultado: 499999500000

--- Suma usando función sum() ---
Tiempo con función sum(): 0.0062 segundos
Resultado: 499999500000

--- Comparación con %timeit ---
Bucle for:
41.8 ms ± 948 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)

Función sum():
5.98 ms ± 24.8 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


In [3]:
# 3. Usando arrays de NumPy

# %%
import numpy as np

# Convertir lista a array numpy
print("--- Usando NumPy ---")
array_np = np.array(lista)
print(f"Array numpy creado con {array_np.size} elementos")

# a) Sumar con bucle for
print("\n--- Suma con bucle for (array numpy) ---")
start_time = time.time()
suma_np_for = 0
for num in array_np:
    suma_np_for += num
end_time = time.time()
print(f"Tiempo con bucle for: {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_np_for}")

# b) Sumar con numpy.sum()
print("\n--- Suma con numpy.sum() ---")
start_time = time.time()
suma_np_func = np.sum(array_np)
end_time = time.time()
print(f"Tiempo con numpy.sum(): {end_time - start_time:.4f} segundos")
print(f"Resultado: {suma_np_func}")

# Comparar con timeit
print("\n--- Comparación con %timeit ---")
print("Bucle for en array numpy:")
%timeit -r 2 -n 1 suma_np_for = 0; [suma_np_for := suma_np_for + num for num in array_np]

print("\nnumpy.sum():")
%timeit -r 2 -n 1 np.sum(array_np)

--- Usando NumPy ---
Array numpy creado con 1000000 elementos

--- Suma con bucle for (array numpy) ---
Tiempo con bucle for: 0.1129 segundos
Resultado: 499999500000

--- Suma con numpy.sum() ---
Tiempo con numpy.sum(): 0.0006 segundos
Resultado: 499999500000

--- Comparación con %timeit ---
Bucle for en array numpy:
63.3 ms ± 150 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)

numpy.sum():
556 μs ± 48.4 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


In [4]:
# 4. Optimización con Numba

from numba import njit

# a) Función original con Numba
@njit
def reduc_operation_numba(a):
    x = 0
    for i in range(a):
        x += i
    return x

# b) Versión con array numpy y numba
@njit
def sum_array_numba(arr):
    x = 0
    for i in range(len(arr)):
        x += arr[i]
    return x

print("=== OPTIMIZACIÓN CON NUMBA ===")

# Prueba con función original
print("\n1. Función original con @njit:")
%timeit -r 2 -n 1 reduc_operation_numba(value)
result_numba = reduc_operation_numba(value)
print(f"Resultado: {result_numba}")

# Prueba con array numpy
print("\n2. Suma de array numpy con @njit (primera ejecución - incluye compilación):")
start_time = time.time()
result_numba_array = sum_array_numba(array_np)
end_time = time.time()
print(f"Tiempo primera ejecución: {end_time - start_time:.4f} segundos")

print("\n3. Suma de array numpy con @njit (ejecuciones subsiguientes):")
%timeit -r 2 -n 1 sum_array_numba(array_np)

# Comparación directa numpy.sum() vs numba
print("\n4. Comparación directa numpy.sum() vs Numba:")

print("numpy.sum():")
%timeit -r 2 -n 1 np.sum(array_np)

print("Numba (@njit):")
%timeit -r 2 -n 1 sum_array_numba(array_np)

=== OPTIMIZACIÓN CON NUMBA ===

1. Función original con @njit:
The slowest run took 67686.68 times longer than the fastest. This could mean that an intermediate result is being cached.
122 ms ± 122 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)
Resultado: 499999500000

2. Suma de array numpy con @njit (primera ejecución - incluye compilación):
Tiempo primera ejecución: 0.0454 segundos

3. Suma de array numpy con @njit (ejecuciones subsiguientes):
292 μs ± 66.6 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)

4. Comparación directa numpy.sum() vs Numba:
numpy.sum():
219 μs ± 27.8 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)
Numba (@njit):
155 μs ± 2.05 μs per loop (mean ± std. dev. of 2 runs, 1 loop each)


# CONCLUSIONES

## 1. Implementación Original
- Tiempo: 38.3 ms (34.3 ms ± 264 μs con timeit)
  
- Esta es la implementación más básica, se trata de un simple bucle for sobre range(). Aunque es sencilla, es relativamente eficiente porque range() genera números sobre la marcha sin crear una estructura de datos completa en memoria. Sin embargo, sigue siendo un bucle en Python puro, que es interpretado y por tanto más lento que código compilado.

## 2. Listas de Python
- Tiempo del bucle for sobre lista: 67.7 ms (42.8 ms ± 1.02 ms con timeit)
- Tiempo de la función sum(): 6.1 ms (6.04 ms ± 5.94 μs con timeit)
  
- El bucle es más lento que la implementación original (67.7 ms vs 38.3 ms). Esto se debe al tiempo que tarda en crear la lista completa en memoria (list(range(10⁶))) y en acceder a los elementos de la lista (más lento que iterar sobre range()).
- Por otro lado, sum() es mucho más rápida que el bucle (6.1 ms vs 67.7 ms) porque está implementada en C (no en Python), evita la sobrecarga del intérprete Python en cada iteración y está optimizada internamente para operaciones de reducción.

## 3. Arrays de NumPy
- Tiempo del bucle for sobre array: 116.5 ms (63.9 ms ± 872 μs con timeit)
- Tiempo de la función numpy.sum(): 0.5 ms (586 μs ± 9.13 μs con timeit)
  
- El bucle sobre array NumPy es 3 veces más lento que la implementación original. Esto ocurre porque los arrays NumPy tienen más sobrecarga para acceso individual, Python debe "desenvolver" cada elemento del array (boxing/unboxing) y se pierden las optimizaciones vectorizadas cuando se usa bucle Python.
- Finalmente, numpy.sum() es el MÁS RÁPIDO, 76 veces más que la implementación original (38.3 ms → 0.5 ms) y 12 veces más rápido que sum() de listas (6.1 ms → 0.5 ms). Esto se debe a que la operación vectorizada se ejecuta en C/Fortran, al uso de SIMD (instrucciones AVX/SSE) del procesador, la optimización de memoria contigua (arrays frente a listas) y la paralelización implícita en algunas operaciones.

## 4. Numba
- Tiempo de Numba (1ª ejecución - incluye compilación): 45.4 ms
- Tiempo de Numba (ejecuciones siguientes, ya compilado): 0.155 ms (155 μs ± 2.05 μs con timeit)
- Tiempo de numpy.sum(): 0.219 ms (219 μs ± 27.8 μs con timeit)

- La primera ejecución de Numba es 1.2 veces más lenta que la implementación original (45.4 ms vs 38.3 ms). Esto se debe al overhead de compilación JIT (Just-In-Time) que Numba realiza durante la primera ejecución para traducir el código Python a instrucciones de máquina optimizadas.
- Una vez compilado, Numba es 247 veces más rápido que la implementación original (0.155 ms vs 38.3 ms) y 1.4 veces más rápido que numpy.sum() (0.155 ms vs 0.219 ms). Esto ocurre porque Numba genera código máquina optimizado específico para la arquitectura del procesador, eliminando la sobrecarga del intérprete de Python.
- Numba supera a NumPy en este caso específico porque la operación es un simple bucle de suma que Numba puede optimizar, mientras que numpy.sum() tiene cierta sobrecarga por ser una función más general que maneja múltiples tipos de datos y opciones.

- En definitiva, para pipelines que se ejecutan repetidamente o algoritmos personalizados con bucles complejos, Numba ofrece la mejor combinación de rendimiento y mantenibilidad. Sin embargo, para operaciones estándar y ejecuciones únicas, NumPy sigue siendo la opción más práctica.

In [5]:
# Para ejecución en terminal 

import sys
import time
import numpy as np
from numba import njit

# Funciones con Numba
@njitsbatch submit_Python_nikola-alumno11.sh 100000000
def reduc_operation_numba(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

@njit
def sum_array_numba(arr):
    """Sum array elements using Numba."""
    x = 0
    for i in range(len(arr)):
        x += arr[i]
    return x

# Función para ejecutar pruebas con un valor de N
def run_tests(N):
    print(f"\n=== EJECUCIÓN CON N = {N:,} ===")
    
    # 1. Original
    start = time.time()
    suma = 0
    for i in range(N):
        suma += i
    t_original = time.time() - start
    
    # 2. Lista + sum()
    lista = list(range(N))
    start = time.time()
    suma_lista = sum(lista)
    t_lista_sum = time.time() - start
    
    # 3. NumPy + np.sum()
    array_np = np.array(lista)
    start = time.time()
    suma_np = np.sum(array_np)
    t_np_sum = time.time() - start
    
    # 4. Numba - función original (incluye compilación)
    start = time.time()
    suma_numba = reduc_operation_numba(N)
    t_numba = time.time() - start
    
    # 5. Numba - función ya compilada
    # Ejecutar otra vez para ver tiempo sin compilación
    start = time.time()
    suma_numba2 = reduc_operation_numba(N)
    t_numba_compilado = time.time() - start
    
    # Resultados
    print(f"\nTiempos de ejecución (segundos):")
    print(f"1. Original (for range):    {t_original:.6f}")
    print(f"2. Lista + sum():           {t_lista_sum:.6f}")
    print(f"3. NumPy + np.sum():        {t_np_sum:.6f}")
    print(f"4. Numba (con compilación): {t_numba:.6f}")
    print(f"5. Numba (ya compilado):    {t_numba_compilado:.6f}")
    
    # Speedups
    if t_lista_sum > 0 and t_np_sum > 0 and t_numba_compilado > 0:
        print(f"\nSpeedups vs Lista+sum():")
        print(f"  NumPy: {t_lista_sum/t_np_sum:.1f}x más rápido")
        print(f"  Numba: {t_lista_sum/t_numba_compilado:.1f}x más rápido")
    
    # Verificación
    if suma == suma_lista == suma_np == suma_numba == suma_numba2:
        print(f"\n✓ Resultado correcto: {suma:,}")
    else:
        print("\n✗ Error: resultados diferentes")
    
    return t_original, t_lista_sum, t_np_sum, t_numba, t_numba_compilado

# Parámetro desde línea de comandos
if len(sys.argv) > 1:
    N = int(sys.argv[1])
    valores = [N]
    print(f"\nParámetro SLURM: N = {N:,}")
else:
    valores = [10**7, 10**8]
    print("\nValores por defecto: 10^7 y 10^8")

# Ejecutar pruebas
resultados = []
for valor in valores:
    tiempos = run_tests(valor)
    resultados.append((valor, *tiempos))
    print("\n" + "-"*60)

# Resumen final
print("\n" + "="*80)
print("RESUMEN FINAL")
print("="*80)
print(f"{'N':<12} {'Original':<10} {'Lista-sum':<10} {'NumPy-sum':<10} {'Numba':<10} {'Numba-comp':<10}")
print("-"*80)

for valor, t_orig, t_lsum, t_nsum, t_numba, t_ncomp in resultados:
    print(f"{valor:<12,} {t_orig:<10.6f} {t_lsum:<10.6f} {t_nsum:<10.6f} {t_numba:<10.6f} {t_ncomp:<10.6f}")

print("="*80)

ValueError: invalid literal for int() with base 10: '-f'

In [13]:
# Código para guardar el notebook
import json
import nbformat as nbf

print("Notebook completo creado.")
print("Guardar como: reduc-operation-alumno11.ipynb")

Notebook completo creado.
Guardar como: reduc-operation-alumno11.ipynb
